### ISSUE TITLE
Implement JWT authentication - Login

### ISSUE THREAD

**Issue by zannis** [SocialHackersClass10/ChalleduApp#46]
As discussed last week, we want to implement JWT authentication for registering our users and providing access to the app's resources

The idea is the following:
Upon registration and login, we issue 2 SHA256-signed tokens and provide them to the user, the JWT access token and the JWT refresh token. These will be created using a secret token, which will be defined in the .env file and signed using the 'jsonwebtoken' npm library.
The access token will be valid for 1 day and the refresh token for 1 week. 
Every request towards the backend that will get or modify server resources will require a valid JWT token attached to the request headers in order to operate. Validity of said tokens will be asserted using an express middleware, as provided by the library 'express-jwt'.
We will NOT implement the complete OAuth2 protocol for our implementation, which is based on JWT tokens as well, but rather a more minimal version of it which will work as follows:

Registration:
Registration is essentially covered by the POST /users endpoint. This endpoint will **NOT** require an authorization header. It will create a user that is pending approval and will **NOT** return any kind of token as part of its response. So no changes needed to the POST /users endpoint as part of this ticket.

Login
We need to implement the login endpoint, which will be a POST endpoint under the /auth/login path. This endpoint will **NOT** require an authorization header and it will receive a JSON object like so:
```
{
    email: <emailaddress>
    password: <the_user_password>
}
```
This endpoint will check mongodb for a user with such an email and will use the bcrypt library to encode it in order to verify whether the provided password matches the one saved in our database. On success, it will return 200 with the following object:
```
{
    access_token: <the JWT access token>
    refresh_token: <the JWT refresh token>
}
```
On failure it will return 401 and the following:
```
{
    error: 'You provided a wrong username or password.'
}
```

Further info and endpoints regarding JWT authentication to be added in the upcoming sprints.

Acceptance criteria:
Add the necessary dependencies to the project
Update the .env file with a secure secret and share it with the rest of the team
Implement the login endpoint

**Comment by student0b4dc0d3** [comment]
created new branch : issue#46-jwt-authentication-login

**Comment by student0b4dc0d3** [comment]
preliminary question #1 
how will a "Pending" user get notified that their registration has been approved, and thus now they can login?

Also, the login endpoint, as described above, has to verify that "document_state" is "Approved", yes?

**Comment by zannis** [comment]
>how will a "Pending" user get notified that their registration has been approved, and thus now they can login?

This is a great question, usually we do this via email notification, however this is a large topic and we will need a third-party email provider service to implement this so I would say that we do the following:
Upon registration, we redirect the user to a screen saying something like: "Thanks for your registration, your account is pending approval, please check again soon for an update." And a link to the login screen. 

> Also, the login endpoint, as described above, has to verify that "document_state" is "Approved", yes?

Yes, in order for the login to succeed, the user has to have document state approved. If verification is pending or rejected, we should just redirect to the respective screen. So this means we would have to implement 2 screens, one saying that your account is still pending approval, and another one saying that your registration has been rejected, please contact XYZ for further info.

I'm open to alternatives, but I don't see what else we could do without adding significant effort to the implementation.

**Comment by student0b4dc0d3** [comment]
About the notification:
Alternatively, we could just send an email to the user with the approval result, like:
* your registration has been approved, you may login anytime with your specified credentials
or
* we would like to inform you that your registration request has been rejected due to XYZ reasons. Please contact XYZ for further info.
That way, we also verify, without any additional effort, the email address.

**Comment by ligopsychakis-michalis** [comment]
Hello team.. If I understand correctly, with the JWT we don't need to search for _id when we trying to get a single user because we use the access token from the header  right?? So we have to change the code for the get single user route???

**Comment by student0b4dc0d3** [comment]
@ligopsychakis-michalis 

erm... no?

The user who is asking for a particular resource is the user that is defined by the access token, and that user may be a completely different entity than the "resource" that is requested by the access request. It may even be an ngo data request, that has nothing to do with "who" is authorised to access a particular resource.

I hope I was clear in expressing a somewhat complicated distinction.
If not, I am always available for more details

Thank you

**Comment by ligopsychakis-michalis** [comment]
Oh you're right @student0b4dc0d3.. Thanks!

**Comment by student0b4dc0d3** [comment]
Does nobody even have a comment for this topic?

For example, do you think we should include the "role" in the payload?
Are there any other considerations you may have?

You shouldn't be afraid to post, or holdback b/c of inexperience, this is part of the learning process.
I promise, I will not bite
,_,

Please leave yer message after the Beep

**Comment by ligopsychakis-michalis** [comment]
thats a good question I think... (about what payload should include)
maybe the login infos are enough ( usename and email ).. these fields are unique right??

**Comment by ligopsychakis-michalis** [comment]
also I want to ask... Is it nesseccery to use the npm packege 'express-jwt'? cause I think we can make the check for token validity through a method of package 'jsonwebtoken'  -> jwt.verify( userToken, ourSecretKey, ( error, user ) => { ... } )  that returns 'error'
it's false or the 'user' if it's true... 
and I think the 'user' will be actually just what the payload includes...

**Comment by zannis** [comment]
Its not necessary to use the express-jwt package, however it will reduce our effort by a great amount. In any case, we want a middleware that exposes the payload of the JWT token as part of the request, so that our handlers will be able to access it.

> Alternatively, we could just send an email to the user with the approval result

To do this via an express app, we need API access to an email provider. There are a lot, and some of them offer free packages, however it is a large dependency that I don't want to get involved with at this point.

To answer the rest of the comments:

The payload **must** contain the user role and some identifier in order for us to be able to find them. Preferably an ID since the email could potentially be 'personal info'.

**Comment by ligopsychakis-michalis** [comment]
ok.. thanks @zannis !

**Comment by student0b4dc0d3** [comment]
We need to implement following parts

* include ACCESS_TOKEN_KEY and REFRESH_TOKEN_KEY in our "dotenv" environmental constants structure

* implement the "auth/login/" endpoint as described above

this endpoint code must :
* ensure the credentials are valid ( existing and belonging to an "Approved" user )
* if found invalid -> report back the error
* if valid -> create both JWT, each with it's own crypto key

JWT payload :
* user "_id" , which is a unique identifier and is used to locate / retrieve the user document
* "role" , this is to fast decide if the **application user** is an admin or not , for fast resource access decision without the need to retrieve user document. This field is included only in the access token
* iat , the token issue date
* xat , the token expiration date

I am, as always, happy to converse on anything. Please don't hesitate to contact me or leave your comment.
Thank you

**Comment by ligopsychakis-michalis** [comment]
Hello team.. so for login route our response are the access and refresh tokens... and the frontend must make a new request to get the user??? but how..? they will not have the _id to get the login user from user/id route.. 
Or we gonna response with tokens and whole login user???

**Comment by student0b4dc0d3** [comment]
Excuse me @ligopsychakis-michalis , I'm not quite sure I completely understood you

Login is the method that will decide to authorize or reject the user defined by provided email & password

If the front-end also absolutely needs us to include the complete user document in the response, we could of course do that.
I am unsure on this tho, since there already exists a route and endpoint to get the user document.

Since you have been involved with that part of the front-end, you may have a more precise oppinion, so what would you suggest is needed ?

**Comment by student0b4dc0d3** [comment]
On second thought, I think I see your point:
we need a way to supply either the user "_id" or the complete user document to the login request, so the front-end can establish a link. You are completely correct on that.
We can do either, or
And since you have been involved in that part of the front-end, I assume your opinion on that will weight the most, so please make your suggestion
Thank you

**Comment by ligopsychakis-michalis** [comment]
hey @student0b4dc0d3.. thanks for the response..
I saw that the frontend takes the fetched data from login route and saves it to their UserContext so they can have access to the login user's infos from all the components.. So I think If everyone agree that we need to response with 
{
    access_token: <the JWT access token>,
    refresh_token: <the JWT refresh token>,
    user: <user>
}

**Comment by student0b4dc0d3** [comment]
Since we will already have retrieved the user document and validated the supplied credentials, I have no objection to that

**Comment by zannis** [comment]
> * iat , the token issue date
> * xat , the token expiration date

@student0b4dc0d3  I completely agree, my only comment is to call the expiration property 'exp' just to be in line with the RFC specification.
https://tools.ietf.org/html/rfc7519

@ligopsychakis-michalis I see your point, but to me, this requirement is covered by adding the _id field inside the payload. The frontend can then use that id to make a request for the user and get all the necessary info. We already have an endpoint for this, so we are essentially ignoring our own design if we start doing 'convenience' changes to our endpoints.

Adding a user object inside our login endpoint is just a 'shortcut' that will save a few lines of code and one API call. But in the long run, it is a bad practice and should be avoided. It might seem 'okay', but consider if you were a third party and used this API without taking part in it's development. When you use the /auth/login endpoint, you would expect to get _just_ the login result information, any other arbitrary data that might be necessary for some implementation will be at least unexpected. This is my rationale behind this, and since there is a very easy way to get past this (by making another API call for the user details) I'd like to keep it that way.

So to sum up, lets stick to the original requirement, and supply only the tokens for the login endpoint. The frontend is responsible to store these tokens and use the _id  contained in the payload to get any further details from the backend API.

**Comment by student0b4dc0d3** [comment]
@zannis 
This was also my 1st thought upon reading @ligopsychakis-michalis 's comment
My 2nd thought , and the one that changed my mind was: although the front-end is supplied with all the "required" info to retrieve the user document, they will have to decode the payload in order to obtain the user _id ( by decode I mean un-base64 , not decrypt )
whereas for us it is very simple to just include the document.

Anyway, we will implement the solution without further user info, and if ever required, we can return to this and include them

**Comment by zannis** [comment]
Lets also update the README with some description for the new env variable please! :)
